renku/core/management/session/docker.py

```diff
@@ -0,0 +1,188 @@
+        )
+
+    @hookimpl
+    def session_provider(self) -> Tuple[ISessionProvider, str]:
+        """Supported session provider.
+
+        :returns: a tuple of ``self`` and provider name.
+        """
+        return (self, "docker")
+
+    @hookimpl
+    def session_list(self, config: Optional[Path], client: LocalClient) -> List[Session]:
+        """Lists all the sessions currently running by the given session provider.
+
+        :returns: a list of sessions.
+        """
+
+        return map(
+            lambda x: Session(
+                id=x.short_id,
+                status=x.status,
+                url=next(DockerSessionProvider()._get_jupyter_urls(x.ports, x.labels["jupyter_token"])),
+            ),
+            self._get_docker_containers(client),
+        )
+
+    @hookimpl
+    def session_start(self, config: Optional[Path], image_name: Optional[str], client: LocalClient) -> str:
+        """Creates an interactive session.
+
+        :returns: a unique id for the created interactive sesssion.
+        """
+
+        def _find_docker_image(remote: Dict[str, str], commit_sha: str) -> docker.models.images.Image:
+            try:
+                return self.docker_client().images.get(DockerSessionProvider._docker_image_name(remote, commit_sha))
+            except docker.errors.ImageNotFound:
+                # try to pull
+                return self.docker_client().images.pull(
+                    DockerSessionProvider._docker_image_name(remote), tag=commit_sha
+                )
+
+        try:
+            docker_is_running = self.docker_client().ping()
+            if not docker_is_running:
+                raise errors.DockerError(
+                    "Could not communicate with the docker instance. Please make sure it is running!"
+                )
+        except docker.api.APIError as error:
+            raise errors.DockerError(error)
+
+        try:
+            remote = client.remote
+            commit_short_sha = client.repository.head.commit.hexsha[:7]
+            if image_name is None:
+                image_name = DockerSessionProvider._docker_image_name(remote, commit_short_sha)
+
+                try:
+                    _ = _find_docker_image(remote, commit_short_sha)
+                except docker.errors.ImageNotFound:
+                    communication.confirm(
+                        f"The docker image '{image_name}' does not exists. Would you like to build it?",
+                        abort=True,
+                    )
+
+                    with yaspin(text="Building docker image"):
+                        _ = self.docker_client().images.build(path=str(client.docker_path.parent), tag=image_name)
+
+            auth_token = uuid4().hex
+            project_config = client.load_config()
+            default_url = project_config.get("interactive", "default_url")
+            container = self.docker_client().containers.run(
+                image_name,
```

Comment:
@vigsterkr it just occurred to me that there you could also have the docker image pinned in the the config in a project. So it may be a good idea to check and use that if it is defined.

It is under `interactive.image`.

Reply:
@olevski do you have a public example somewhere on renkulab that has that set?

Reply:
Also if this complicates things I think it is ok to address it in a followup PR/improvement. I know there is a lot of discussions on how to handle images/building/etc.

Reply:
i mean if the renkulab interactive session is following that logic, then so should the local docker based one.

Reply:
Yeah here: https://dev.renku.ch/projects/tasko.olevski/test-pinned-image-1

Reply:
@olevski wait just to be clear: in this case when you set `image = renku/renkulab-py:3.9-0.10.3` in the `renku.ini` this will set the `RENKU_BASE_IMAGE` or this is purely setting the image that is going to be used in the interactive session. just because if the latter, then all the dependencies that are defined in `requirements.txt` of the given project is gonna be ignored.

Reply:
@vigsterkr yeah if the image is pinned (i.e. if the image field is set in renku.ini) then the dockerfile does not have to be built. When you launch a session on renkulab.io with a pinned image the notebook service immediately tries to pull the pinned image and if that is not available it fails. Similarly if the image is pinned the UI does not check if any of the CI pipelines for the project (that usually build the images) have completed or not. It is assumed that the pinned image is ready to go. So yeah with a pinned image all the stuff in `requirements.txt` and even in the Dockerfile is ignored.